File: apps/proxy/internal/providers/perch_stream.go
package providers

import (
	"bufio"
	"encoding/json"
	"io"
	"net/http"
	"strings"
	"time"
)

// perchSSEToChatSSEReader translates the Perch model-call SSE stream into an
// OpenAI chat.completion.chunk SSE stream. Perch events mirror the CLI wire
// format: reasoning_delta / answer_delta text deltas, tool_call_delta and
// tool_use_end tool-call events, and a terminal done event carrying usage.
func perchSSEToChatSSEReader(source io.Reader, model string, includeReasoning bool) io.Reader {
	reader, writer := io.Pipe()
	go func() {
		transformPerchSSEToChat(source, writer, model, includeReasoning)
		_ = writer.Close()
	}()
	return reader
}

type perchStreamTool struct {
	index       int
	name        string
	emittedArgs bool
}

func perchWriteChunk(writer io.Writer, completionID, model string, delta map[string]any, finish any, usage map[string]any) {
	chunk := map[string]any{
		"id":      completionID,
		"object":  "chat.completion.chunk",
		"created": time.Now().Unix(),
		"model":   model,
		"choices": []any{map[string]any{"index": 0, "delta": delta, "finish_reason": finish}},
	}
	if usage != nil {
		chunk["usage"] = usage
	}
	encoded, _ := json.Marshal(chunk)
	_, _ = writer.Write([]byte("data: " + string(encoded) + "\n\n"))
}

func perchToolDeltaChunk(writer io.Writer, completionID, model string, index int, withID bool, id string, withName bool, name string, arguments string) {
	toolDelta := map[string]any{"index": index}
	function := map[string]any{}
	if withID {
		toolDelta["id"] = id
		toolDelta["type"] = "function"
	}
	if withName {
		function["name"] = name
	}
	if arguments != "" {
		function["arguments"] = arguments
	}
	toolDelta["function"] = function
	perchWriteChunk(writer, completionID, model, map[string]any{"tool_calls": []any{toolDelta}}, nil, nil)
}

func transformPerchSSEToChat(source io.Reader, writer io.Writer, model string, includeReasoning bool) {
	completionID := randomID("chatcmpl")
	sentRole := false
	nextToolIndex := 0
	tools := map[string]*perchStreamTool{}

	scanner := bufio.NewScanner(source)
	scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" || line == "[DONE]" || strings.HasPrefix(line, ":") {
			continue
		}
		payload := strings.TrimSpace(strings.TrimPrefix(line, "data:"))
		if payload == "" || payload == "[DONE]" {
			continue
		}
		var event map[string]any
		if err := json.Unmarshal([]byte(payload), &event); err != nil {
			continue
		}
		switch stringValue(event["type"]) {
		case "reasoning_delta":
			if !includeReasoning {
				continue
			}
			perchEnsureRole(writer, completionID, model, &sentRole)
			if delta := stringValue(event["text"]); delta != "" {
				perchWriteChunk(writer, completionID, model, map[string]any{"reasoning_content": delta}, nil, nil)
			}
		case "answer_delta":
			perchEnsureRole(writer, completionID, model, &sentRole)
			if delta := stringValue(event["text"]); delta != "" {
				perchWriteChunk(writer, completionID, model, map[string]any{"content": delta}, nil, nil)
			}
		case "tool_call_delta", "tool_use_end":
			sealed := stringValue(event["type"]) == "tool_use_end"
			for _, rawCall := range perchEventToolCalls(event) {
				call, _ := rawCall.(map[string]any)
				id := stringValue(call["id"])
				if id == "" {
					continue
				}
				state, ok := tools[id]
				if !ok {
					perchEnsureRole(writer, completionID, model, &sentRole)
					name := stringValue(call["name"])
					state = &perchStreamTool{index: nextToolIndex, name: name}
					tools[id] = state
					nextToolIndex++
					perchToolDeltaChunk(writer, completionID, model, state.index, true, id, name != "", name, "")
				}
				if name := stringValue(call["name"]); name != "" && state.name == "" {
					state.name = name
					perchToolDeltaChunk(writer, completionID, model, state.index, false, "", true, name, "")
				}
				arguments := perchToolArgumentsDelta(call, sealed)
				if arguments == "" {
					continue
				}
				if sealed && state.emittedArgs {
					continue
				}
				state.emittedArgs = true
				perchToolDeltaChunk(writer, completionID, model, state.index, false, "", false, "", arguments)
			}
		case "done":
			ok, _ := event["ok"].(bool)
			if !ok {
				if message := perchErrorMessage(event); message != "" {
					perchEnsureRole(writer, completionID, model, &sentRole)
					perchWriteChunk(writer, completionID, model, map[string]any{"content": message}, nil, nil)
				}
			}
			finish := "stop"
			if len(tools) > 0 && ok {
				finish = "tool_calls"
			}
			perchWriteChunk(writer, completionID, model, map[string]any{}, finish, perchUsageToChatUsage(event["usage"]))
			_, _ = writer.Write([]byte("data: [DONE]\n\n"))
			return
		default:
			continue
		}
	}
	perchWriteChunk(writer, completionID, model, map[string]any{}, "stop", nil)
	_, _ = writer.Write([]byte("data: [DONE]\n\n"))
}

func perchEnsureRole(writer io.Writer, completionID, model string, sentRole *bool) {
	if *sentRole {
		return
	}
	perchWriteChunk(writer, completionID, model, map[string]any{"role": "assistant", "content": ""}, nil, nil)
	*sentRole = true
}

// perchSSEToChatCompletion buffers a full Perch SSE stream and synthesizes a
// single OpenAI chat.completion object for non-streaming requests.
func perchSSEToChatCompletion(source io.Reader, model string, includeReasoning bool) (map[string]any, error) {
	var content, reasoning strings.Builder
	toolCalls := []any{}
	finishReason := "stop"
	var usage map[string]any

	type bufferedTool struct {
		id   string
		name string
		args strings.Builder
	}
	var orderedTools []*bufferedTool
	byID := map[string]*bufferedTool{}
	ensureTool := func(id, name string) *bufferedTool {
		if tool, ok := byID[id]; ok {
			if name != "" && tool.name == "" {
				tool.name = name
			}
			return tool
		}
		tool := &bufferedTool{id: id, name: name}
		byID[id] = tool
		orderedTools = append(orderedTools, tool)
		return tool
	}

	scanner := bufio.NewScanner(source)
	scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" || line == "[DONE]" || strings.HasPrefix(line, ":") {
			continue
		}
		payload := strings.TrimSpace(strings.TrimPrefix(line, "data:"))
		if payload == "" || payload == "[DONE]" {
			continue
		}
		var event map[string]any
		if err := json.Unmarshal([]byte(payload), &event); err != nil {
			continue
		}
		switch stringValue(event["type"]) {
		case "reasoning_delta":
			if includeReasoning {
				if delta := stringValue(event["text"]); delta != "" {
					reasoning.WriteString(delta)
				}
			}
		case "answer_delta":
			if delta := stringValue(event["text"]); delta != "" {
				content.WriteString(delta)
			}
		case "tool_call_delta", "tool_use_end":
			sealed := stringValue(event["type"]) == "tool_use_end"
			for _, rawCall := range perchEventToolCalls(event) {
				call, _ := rawCall.(map[string]any)
				id := stringValue(call["id"])
				if id == "" {
					continue
				}
				tool := ensureTool(id, stringValue(call["name"]))
				if sealed {
					if arguments := perchToolSealedArguments(call); arguments != "" {
						tool.args.Reset()
						tool.args.WriteString(arguments)
					}
				} else if delta := stringValue(call["rawArgumentsText"]); delta != "" {
					tool.args.WriteString(delta)
				}
			}
		case "done":
			message := perchErrorMessage(event)
			ok, hasOK := event["ok"].(bool)
			if message != "" || (hasOK && !ok) {
				if message == "" {
					message = "Perch request failed"
				}
				return nil, &perchUpstreamError{Message: message, Quota: perchQuotaError(message)}
			}
			usage = perchUsageToChatUsage(event["usage"])
		}
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}

	for _, tool := range orderedTools {
		arguments := tool.args.String()
		if strings.TrimSpace(arguments) == "" {
			arguments = "{}"
		}
		toolCalls = append(toolCalls, map[string]any{
			"id":       tool.id,
			"type":     "function",
			"function": map[string]any{"name": tool.name, "arguments": arguments},
		})
	}

	message := map[string]any{"role": "assistant", "content": nil}
	if text := content.String(); text != "" {
		message["content"] = text
	}
	if includeReasoning {
		if text := reasoning.String(); text != "" {
			message["reasoning_content"] = text
		}
	}
	if len(toolCalls) > 0 {
		message["tool_calls"] = toolCalls
		finishReason = "tool_calls"
	}
	if usage == nil {
		usage = map[string]any{"prompt_tokens": 0, "completion_tokens": 0, "total_tokens": 0}
	}

	return map[string]any{
		"id":      randomID("chatcmpl"),
		"object":  "chat.completion",
		"created": time.Now().Unix(),
		"model":   model,
		"choices": []any{map[string]any{"index": 0, "message": message, "finish_reason": finishReason}},
		"usage":   usage,
	}, nil
}

func perchEventToolCalls(event map[string]any) []any {
	if value, ok := event["toolCalls"].([]any); ok {
		return value
	}
	if value, ok := event["tool_calls"].([]any); ok {
		return value
	}
	return nil
}

// perchToolArgumentsDelta returns the argument text for a streaming chunk:
// raw streamed text while open, and the sealed JSON arguments object at close.
func perchToolArgumentsDelta(call map[string]any, sealed bool) string {
	if raw := stringValue(call["rawArgumentsText"]); raw != "" {
		return raw
	}
	if sealed {
		return perchToolSealedArguments(call)
	}
	return ""
}

func perchToolSealedArguments(call map[string]any) string {
	if text := stringValue(call["arguments"]); text != "" {
		return text
	}
	if arguments := call["arguments"]; arguments != nil {
		encoded, _ := json.Marshal(arguments)
		return string(encoded)
	}
	return ""
}

func perchErrorMessage(event map[string]any) string {
	if text := stringValue(event["error"]); text != "" {
		return text
	}
	if value := event["error"]; value != nil {
		encoded, _ := json.Marshal(value)
		return string(encoded)
	}
	return ""
}

type perchUpstreamError struct {
	Message string
	Quota   bool
}

func (e *perchUpstreamError) Error() string {
	return e.Message
}

func perchErrorStatus(err *perchUpstreamError) int {
	if err.Quota {
		return http.StatusTooManyRequests
	}
	return http.StatusBadGateway
}

func perchErrorType(err *perchUpstreamError) string {
	if err.Quota {
		return "rate_limit_error"
	}
	return "api_error"
}

func perchQuotaError(message string) bool {
	lower := strings.ToLower(message)
	for _, marker := range []string{"allowance", "quota", "limit", "usage", "billing", "credit"} {
		if strings.Contains(lower, marker) {
			return true
		}
	}
	return false
}

func perchUsageToChatUsage(raw any) map[string]any {
	usage, _ := raw.(map[string]any)
	if usage == nil {
		return nil
	}
	input := perchUsageInt(usage, "inputTokens")
	output := perchUsageInt(usage, "outputTokens")
	cacheRead := perchUsageInt(usage, "cacheReadInputTokens")
	if input == 0 && output == 0 && cacheRead == 0 {
		return nil
	}
	promptTokens := input + cacheRead
	return map[string]any{"prompt_tokens": promptTokens, "completion_tokens": output, "total_tokens": promptTokens + output}
}

func perchUsageInt(usage map[string]any, key string) int {
	if value, ok := usage[key].(float64); ok && value > 0 {
		return int(value)
	}
	return 0
}

File: packages/ai/src/protocol/responses/perch-stream.ts
export interface PerchStreamCallbacks {
    onChunk(chunk: Record<string, unknown>): void;
    onDone(): void;
}

interface PerchToolCall {
    id?: unknown;
    name?: unknown;
    rawArgumentsText?: unknown;
    arguments?: unknown;
}

interface PerchEvent {
    type?: unknown;
    text?: unknown;
    toolCalls?: unknown;
    tool_calls?: unknown;
    ok?: unknown;
    error?: unknown;
    usage?: unknown;
}

interface ChatUsage {
    prompt_tokens: number;
    completion_tokens: number;
    total_tokens: number;
}

export class PerchUpstreamError extends Error {
    readonly quota: boolean;

    constructor(message: string) {
        super(message);
        this.name = "PerchUpstreamError";
        this.quota = isQuotaError(message);
    }
}

function randomCompletionId(): string {
    const suffix =
        globalThis.crypto?.randomUUID?.().replaceAll("-", "") ??
        Math.random().toString(36).slice(2);
    return `chatcmpl_${suffix}`;
}

function stringValue(value: unknown): string {
    return typeof value === "string" ? value : "";
}

function eventToolCalls(event: PerchEvent): PerchToolCall[] {
    const calls = Array.isArray(event.toolCalls)
        ? event.toolCalls
        : event.tool_calls;
    return Array.isArray(calls)
        ? calls.filter(
              (call): call is PerchToolCall =>
                  Boolean(call) && typeof call === "object",
          )
        : [];
}

function sealedArguments(call: PerchToolCall): string {
    if (typeof call.arguments === "string") return call.arguments;
    if (call.arguments === undefined || call.arguments === null) return "";
    try {
        return JSON.stringify(call.arguments);
    } catch {
        return "";
    }
}

function errorMessage(event: PerchEvent): string {
    if (typeof event.error === "string") return event.error;
    if (event.error === undefined || event.error === null) return "";
    try {
        return JSON.stringify(event.error);
    } catch {
        return String(event.error);
    }
}

function isQuotaError(message: string): boolean {
    const lower = message.toLowerCase();
    return ["allowance", "quota", "limit", "usage", "billing", "credit"].some(
        (marker) => lower.includes(marker),
    );
}

function usageToChatUsage(raw: unknown): ChatUsage | null {
    if (!raw || typeof raw !== "object") return null;
    const usage = raw as Record<string, unknown>;
    const positiveInteger = (key: string): number => {
        const value = usage[key];
        return typeof value === "number" && Number.isFinite(value) && value > 0
            ? Math.trunc(value)
            : 0;
    };
    const input = positiveInteger("inputTokens");
    const output = positiveInteger("outputTokens");
    const cacheRead = positiveInteger("cacheReadInputTokens");
    if (input === 0 && output === 0 && cacheRead === 0) return null;
    const prompt = input + cacheRead;
    return {
        prompt_tokens: prompt,
        completion_tokens: output,
        total_tokens: prompt + output,
    };
}

export function createPerchSSEToChatTransformer(
    model: string,
    callbacks: PerchStreamCallbacks,
    includeReasoning = true,
) {
    const completionId = randomCompletionId();
    let sentRole = false;
    let nextToolIndex = 0;
    let finished = false;
    const tools = new Map<
        string,
        { index: number; name: string; emittedArgs: boolean }
    >();

    function writeChunk(
        delta: Record<string, unknown>,
        finishReason: string | null = null,
        usage: ChatUsage | null = null,
    ) {
        const chunk: Record<string, unknown> = {
            id: completionId,
            object: "chat.completion.chunk",
            created: Math.floor(Date.now() / 1000),
            model,
            choices: [{ index: 0, delta, finish_reason: finishReason }],
        };
        if (usage) chunk.usage = usage;
        callbacks.onChunk(chunk);
    }

    function ensureRole() {
        if (sentRole) return;
        writeChunk({ role: "assistant", content: "" });
        sentRole = true;
    }

    function finish(finishReason = "stop", usage: ChatUsage | null = null) {
        if (finished) return;
        finished = true;
        writeChunk({}, finishReason, usage);
        callbacks.onDone();
    }

    return {
        processEvent(rawEvent: unknown) {
            if (finished || !rawEvent || typeof rawEvent !== "object") return;
            const event = rawEvent as PerchEvent;
            const type = stringValue(event.type);

            if (type === "reasoning_delta") {
                if (!includeReasoning) return;
                ensureRole();
                const text = stringValue(event.text);
                if (text) writeChunk({ reasoning_content: text });
                return;
            }

            if (type === "answer_delta") {
                ensureRole();
                const text = stringValue(event.text);
                if (text) writeChunk({ content: text });
                return;
            }

            if (type === "tool_call_delta" || type === "tool_use_end") {
                const sealed = type === "tool_use_end";
                for (const call of eventToolCalls(event)) {
                    const id = stringValue(call.id);
                    if (!id) continue;

                    let state = tools.get(id);
                    if (!state) {
                        ensureRole();
                        const name = stringValue(call.name);
                        state = {
                            index: nextToolIndex++,
                            name,
                            emittedArgs: false,
                        };
                        tools.set(id, state);
                        writeChunk({
                            tool_calls: [
                                {
                                    index: state.index,
                                    id,
                                    type: "function",
                                    function: name ? { name } : {},
                                },
                            ],
                        });
                    }

                    const lateName = stringValue(call.name);
                    if (lateName && !state.name) {
                        state.name = lateName;
                        writeChunk({
                            tool_calls: [
                                {
                                    index: state.index,
                                    function: { name: lateName },
                                },
                            ],
                        });
                    }

                    const rawArguments = stringValue(call.rawArgumentsText);
                    const args =
                        rawArguments || (sealed ? sealedArguments(call) : "");
                    if (!args || (sealed && state.emittedArgs)) continue;
                    state.emittedArgs = true;
                    writeChunk({
                        tool_calls: [
                            {
                                index: state.index,
                                function: { arguments: args },
                            },
                        ],
                    });
                }
                return;
            }

            if (type === "done") {
                const ok = event.ok === true;
                if (!ok) {
                    const message = errorMessage(event);
                    if (message) {
                        ensureRole();
                        writeChunk({ content: message });
                    }
                }
                finish(
                    tools.size > 0 && ok ? "tool_calls" : "stop",
                    usageToChatUsage(event.usage),
                );
            }
        },
        finish() {
            finish();
        },
    };
}

export async function readPerchSSE(
    source: ReadableStream<Uint8Array>,
    onEvent: (event: unknown) => void,
): Promise<void> {
    const reader = source.getReader();
    const decoder = new TextDecoder();
    let buffer = "";

    const processBlock = (block: string) => {
        const data = block
            .split(/\r?\n/)
            .filter((line) => line.startsWith("data:"))
            .map((line) => line.slice(5).trimStart())
            .join("\n")
            .trim();
        if (!data || data === "[DONE]") return;
        try {
            onEvent(JSON.parse(data));
        } catch {
            // Perch occasionally emits keep-alives and non-JSON diagnostics.
        }
    };

    try {
        while (true) {
            const { done, value } = await reader.read();
            if (done) break;
            buffer += decoder.decode(value, { stream: true });
            const blocks = buffer.split(/\r?\n\r?\n/);
            buffer = blocks.pop() ?? "";
            for (const block of blocks) processBlock(block);
        }
        buffer += decoder.decode();
        if (buffer.trim()) processBlock(buffer);
    } finally {
        reader.releaseLock();
    }
}

export async function perchSSEToChatCompletion(
    source: ReadableStream<Uint8Array>,
    model: string,
    includeReasoning = true,
): Promise<Record<string, unknown>> {
    let content = "";
    let reasoning = "";
    let usage: ChatUsage | null = null;
    let upstreamError: PerchUpstreamError | null = null;
    const tools: Array<{ id: string; name: string; arguments: string }> = [];
    const toolsById = new Map<string, (typeof tools)[number]>();

    await readPerchSSE(source, (rawEvent) => {
        if (!rawEvent || typeof rawEvent !== "object") return;
        const event = rawEvent as PerchEvent;
        const type = stringValue(event.type);
        if (type === "reasoning_delta" && includeReasoning) {
            reasoning += stringValue(event.text);
        } else if (type === "answer_delta") {
            content += stringValue(event.text);
        } else if (type === "tool_call_delta" || type === "tool_use_end") {
            const sealed = type === "tool_use_end";
            for (const call of eventToolCalls(event)) {
                const id = stringValue(call.id);
                if (!id) continue;
                let tool = toolsById.get(id);
                if (!tool) {
                    tool = { id, name: stringValue(call.name), arguments: "" };
                    toolsById.set(id, tool);
                    tools.push(tool);
                } else if (!tool.name) {
                    tool.name = stringValue(call.name);
                }
                if (sealed) {
                    const args = sealedArguments(call);
                    if (args) tool.arguments = args;
                } else {
                    tool.arguments += stringValue(call.rawArgumentsText);
                }
            }
        } else if (type === "done") {
            const message = errorMessage(event);
            if (message || event.ok === false) {
                upstreamError = new PerchUpstreamError(
                    message || "Perch request failed",
                );
            }
            usage = usageToChatUsage(event.usage);
        }
    });

    if (upstreamError) throw upstreamError;

    const message: Record<string, unknown> = {
        role: "assistant",
        content: content || null,
    };
    if (includeReasoning && reasoning) message.reasoning_content = reasoning;
    if (tools.length > 0) {
        message.tool_calls = tools.map((tool) => ({
            id: tool.id,
            type: "function",
            function: {
                name: tool.name,
                arguments: tool.arguments.trim() ? tool.arguments : "{}",
            },
        }));
    }

    return {
        id: randomCompletionId(),
        object: "chat.completion",
        created: Math.floor(Date.now() / 1000),
        model,
        choices: [
            {
                index: 0,
                message,
                finish_reason: tools.length > 0 ? "tool_calls" : "stop",
            },
        ],
        usage: usage ?? {
            prompt_tokens: 0,
            completion_tokens: 0,
            total_tokens: 0,
        },
    };
}
